### PR TITLE
fix: Run all audio streaming process on the audio thread or the worker thread

### PR DESCRIPTION
**Pull request by karasusan**
Related issues #545, #525, #602
First, this pull request is affected from @aet suggestion [here](#613). So, he is a co-author of this pull request.
I appreciate his great contribution.

Unity provides the way of implementation to process something on audio thread. [`MonoBehaviour.OnAudioFilterRead(float[], int)`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnAudioFilterRead.html) API is a callback which is called from the audio thread, and we can read/write the audio buffer before playing audio.

The basic idea is running tasks of audio streaming on the audio thread instead of the main thread. For this fix, playing audio smoothly even if the main thread has high load.

Additionally, we adopted the pull model for audio handling. The detail of design is described on the @aet 's pull request (#613).

### CODE REVIEW

**Comment by aet** [comment]
I did some testing, sounds good to me! Thanks for cleaning up my old C style idioms from the code and simplifying it further, the original use case was a bit different. The audio render callback was actually calling transport_->NeedMorePlayData directly and you didn't necessarily get enough samples, the ringbuffer read/write all happened in the same thread. Writing the leftovers into ringbuffer etc.

I'm fine leaving OnAudioReceived alive, though I personally don't see where it could be useful :)

Some minor cleanups at https://github.com/aet/com.unity.webrtc/commit/09235adcf67984c56962188751af016e74cf8fc6 even if the callback only happens once, I'd prefer not dealing with the mutex before it's done

**Comment by karasusan** [comment]
@aet 
Thanks for your review!

As you said, `OnAudioReceived` callback is not needed.
I fixed samples that depending the callback and I removed it.

I am experimenting that keeping shorter the ring buffer size.

**Comment by aet** [comment]
One thing I didn't test was AudioStreamTrack(AudioSource source) code path. I see AudioSourceRead was removed but I'm not sure if AudioCustomFilter can replace it as-is? I was more interested getting rid of AudioClips elsewhere :) If it currently doesn't work as expected, maybe it can be changed to use a different variant of the filter and still rely on OnAudioFilterRead

File: Runtime/Scripts/AudioStreamTrack.cs [hunk 1]
```diff
@@ -52,133 +52,70 @@ public class AudioStreamTrack : MediaStreamTrack
...
-                int lengthSamples = sampleRate;  // sample length for 1 second
-                return AudioClip.Create(
-                    clipName, lengthSamples, channels, sampleRate, false);
+                if (_filter != null)
+                    return;
+                _filter = GetOrAddComponent<AudioCustomFilter>(source.gameObject);
+                _filter.hideFlags = HideFlags.HideInInspector;
+                _filter.onAudioRead += SetData;
+                source.Play();
+                Debug.Log("AddFilter");
```
Comment (kannan-xiao4) [inline Runtime/Scripts/AudioStreamTrack.cs:114]: I think no need this log in package script

File: Plugin~/WebRTCPlugin/WebRTCPlugin.h [hunk 2]
```diff
@@ -25,12 +25,7 @@ namespace webrtc
     using DelegateSetSessionDescFailure =
         void(*)(PeerConnectionObject*, webrtc::RTCErrorType, const char*);
     using DelegateAudioReceive =
-        void(*)(AudioTrackSinkAdapter* sink,
-            const void* audio_data,
-            int size,
-            int sample_rate,
-            int number_of_channels,
-            int number_of_frames);
+        void(*)(AudioTrackSinkAdapter* sink);
```
Comment (kannan-xiao4) [inline Plugin~/WebRTCPlugin/WebRTCPlugin.h:28]: I think this definition is no needed.

File: Runtime/Scripts/AudioStreamTrack.cs (hunk 1 above)
Comment (karasusan) [reply Runtime/Scripts/AudioStreamTrack.cs]: fixed it

File: Plugin~/WebRTCPlugin/WebRTCPlugin.h (hunk 2 above)
Comment (karasusan) [reply Plugin~/WebRTCPlugin/WebRTCPlugin.h]: fixed it

**Review by kannan-xiao4 — APPROVED** [review]

**Review by kannan-xiao4 — APPROVED** [review]

**Comment by karasusan** [comment]
@aet AudioStreamTrack(AudioSource source) constructor is used by some samples, so I tested them but there is no problem.

File: Runtime/Scripts/AudioStreamTrack.cs [hunk 3]
```diff
@@ -319,23 +186,23 @@ public AudioStreamTrack(AudioSource source)
                 throw new ArgumentNullException("AudioSource argument is null.");
             if (source.clip == null)
```
Comment (karasusan) [inline Runtime/Scripts/AudioStreamTrack.cs:187]: @aet 
I may have misunderstood your comment.
Did you mean we can remove this line?

Comment (aet) [reply Runtime/Scripts/AudioStreamTrack.cs:187]: No misunderstanding, glad to hear AudioCustomFilter works for both cases. I only realized later it was using a different SetData implementation!

I had the clip check removed before, but thought maybe it was there for a reason. I'm all forward removing it if no harm done

Comment (karasusan) [reply Runtime/Scripts/AudioStreamTrack.cs:187]: I will make another PR for this fix.

**Comment by aet** [comment]
Doesn't require updating the plugins, but noticed that Plugin~/WebRTCPlugin/AudioTrackSinkAdapter.cpp now includes common_audio/include/audio_util.h twice. easy to clean up at any point though

**Comment by aet** [comment]
One more cleanup, unused public delegate void OnAudioReceived(AudioSource renderer) declared in Runtime/Scripts/AudioStreamTrack.cs

**Pull request merged**